quickwit/quickwit-indexing/src/source/pulsar_source.rs

```diff
@@ -0,0 +1,1083 @@
+        }
+    }
+
+    fn push(&mut self, doc: String, num_bytes: u64) {
+        self.docs.push(doc);
+        self.num_bytes += num_bytes;
+    }
+}
+
+struct PulsarConsumer {
+    messages: mpsc::Receiver<ConsumerMessage>,
+    last_ack: watch::Sender<SourceCheckpoint>,
+}
+
+#[tracing::instrument(name = "pulsar-consumer", skip(pulsar))]
+/// Spawns a background task listening for incoming messages to
+/// process.
+///
+/// The exists because the `Consumer` is not `Sync` or `Send` in places which means we cannot
+/// have it as part of the `Source` impl itself or as a background tokio task.
+///
+/// Instead we spawn a separate tokio runtime and use a local set to ensure the bounds.
+async fn spawn_message_listener(
+    subscription_name: String,
+    params: PulsarSourceParams,
+    pulsar: Pulsar<TokioExecutor>,
+    previous_positions: BTreeMap<PartitionId, Position>,
+) -> anyhow::Result<PulsarConsumer> {
+    let (messages_tx, messages_rx) = mpsc::channel(100);
+    let (last_ack_tx, last_ack_rx) = watch::channel(SourceCheckpoint::default());
+    let (ready, waiter) = oneshot::channel();
+
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()?;
+
+    let handle = std::thread::spawn(move || {
+        let set = LocalSet::new();
+        let fut = set.run_until(async move {
+            let mut consumer: Consumer<PulsarMessage, _> = pulsar
+                .consumer()
+                .with_topics(&params.topics)
+                .with_consumer_name(&params.consumer_name)
+                .with_subscription(subscription_name)
+                .with_subscription_type(SubType::Failover)
+                .build()
+                .await?;
+
+            let consumer_ids = consumer.consumer_id()
+                .into_iter()
+                .map(|id| id.to_string())
+                .collect::<Vec<_>>();
+            info!(positions = ?previous_positions, "Seeking to last checkpoint position.");
+            for (_, position) in previous_positions {
+                let seek_to = msg_id_from_position(&position);
+
+                if seek_to.is_some() {
+                    consumer.seek(Some(consumer_ids.clone()), seek_to, None, pulsar.clone()).await?;
+                }
+            }
+
+            let _ = ready.send(());
+
+            'outer: loop {
+                while let Ok(msg_opt) = timeout(MESSAGE_WAIT_TIMEOUT, consumer.next()).await {
+                    match msg_opt {
+                        None => break 'outer,
+                        Some(msg) => {
+                            if messages_tx.send(msg).await.is_err() {
+                                break 'outer;
+                            }
+                        }
+                    };
+                }
+
+                if last_ack_rx.has_changed().unwrap_or_default() {
+                    let checkpoint = last_ack_rx.borrow();
+
+                    for (partition, position) in checkpoint.iter() {
```

Comment:
I think we must be a bit more clearer on what is a partition here.
Here a partition is `{topic}-{partitionId}`. So we should extract the partition ID from it. Storing the partition ID inside the position is redundant and strange from a external viewpoint (we have a btree map of partition ID -> Position).

Reply:
I think we discussed this last week, but ultimately it's better that we do not try and connect these two things, it's better for them to be separate to avoid confusion and in order to be correct.